sound/soc/sof/ipc4-pcm.c

```diff
@@ -413,7 +413,19 @@ static int sof_ipc4_trigger_pipelines(struct snd_soc_component *component,
 	ret = sof_ipc4_set_multi_pipeline_state(sdev, state, trigger_list);
 	if (ret < 0) {
 		dev_err(sdev->dev, "failed to set final state %d for all pipelines\n", state);
-		goto free;
+		/*
+		 * workaround: if the firmware crashes during stop on pipeline
+		 * reset we must continue cleaning up the state by clearing the
+		 * return error code.
+		 * Since the firmware is crashed we will not send IPC messages
+		 * and we are going to see errors printed, but the state of the
+		 * widgets will be correct for the next boot.
+		 */
+		if (sdev->fw_state != SOF_FW_CRASHED ||
+		    !(cmd == SNDRV_PCM_TRIGGER_STOP && state == SOF_IPC4_PIPE_RESET))
```

Comment:
Checking only for `state == SOF_IPC4_PIPE_RESET` is enough for the workaround.

I'll update the patch and the commit message for more details I have gathered since yesterday.

Reply:
@ujfalusi can we do something like if (state == RESET && ret != -ETIMEDOUT) instead?

Reply:
if the firmware crashed before (during audio playback/capture) then the error is not going to be timeout, but -ENODEV and we need to ignore that as well.